<h3>Load Data

In [1]:
import numpy as np
import pandas as pd

df = pd.read_csv('dataset/voice.csv', encoding='latin-1') # spesifiksi encoding diperlukan karena data tidak menggunakan UTF-8

df.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

### 1. Buatlah model klasifikasi dengan menggunakan SVM untuk data suara, voice.csv. 

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

# Membaca data dari file CSV
dpath = 'dataset/voice.csv'
data = pd.read_csv(dpath, encoding='latin-1') # spesifiksi encoding diperlukan karena data tidak menggunakan UTF-8
data.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

In [3]:
# Pisahkan atribut (X) dan label (y)
X = data.iloc[:, :-1]
y = data.iloc[:, -1]

# Menggunakan LabelEncoder untuk mengubah label menjadi numerik
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)

# Bagi data menjadi data pelatihan dan data pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [4]:
# Buat model SVM
svm_classifier = SVC(kernel='linear', C=1.0, random_state=42)

# Latih model dengan data pelatihan
svm_classifier.fit(X_train, y_train)


SVC(kernel='linear', random_state=42)

In [5]:
# Lakukan prediksi pada data pengujian
y_pred = svm_classifier.predict(X_test)

In [8]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
# Cetak laporan klasifikasi
print(classification_report(y_test, y_pred))

# Cetak matriks kebingungan
print(confusion_matrix(y_test, y_pred))

# Hitung akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f'Akurasi: {accuracy}')
print(f'Akurasi (%): {accuracy * 100:.2f}%')

              precision    recall  f1-score   support

           0       0.98      0.89      0.93       297
           1       0.91      0.99      0.95       337

    accuracy                           0.94       634
   macro avg       0.95      0.94      0.94       634
weighted avg       0.94      0.94      0.94       634

[[265  32]
 [  5 332]]
Akurasi: 0.9416403785488959
Akurasi (%): 94.16%


### 2. Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,
- Menggunakan data spam.csv
- Fitur CountVectorizer dengan mengaktifkan stop_words
- Evaluasi hasilnya

In [10]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score


# Membaca data dari file CSV
dpath = 'dataset/spam.csv'
data = pd.read_csv(dpath, encoding='latin-1') # spesifiksi encoding diperlukan karena data tidak menggunakan UTF-8
data.head()

# Pisahkan atribut (X) dan label (y)
X = data['v2']
y = data['v1']

# Bagi data menjadi data pelatihan dan data pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Buat CountVectorizer dengan mengaktifkan stop words:

In [11]:
from sklearn.feature_extraction.text import CountVectorizer

# Buat objek CountVectorizer dengan pengaktifan stop_words
vectorizer = CountVectorizer(stop_words='english')

# Transformasikan data teks pelatihan dan pengujian menjadi matriks fitur
X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec = vectorizer.transform(X_test)

Buat model Multinomial Naive Bayes dan latih model:

In [12]:
from sklearn.naive_bayes import MultinomialNB

# Buat model Multinomial Naive Bayes
nb_classifier = MultinomialNB()

# Latih model dengan data pelatihan yang telah di-CountVectorized
nb_classifier.fit(X_train_vec, y_train)

MultinomialNB()

Lakukan prediksi dengan model yang telah dilatih:

In [13]:
# Lakukan prediksi pada data pengujian yang telah di-CountVectorized
y_pred = nb_classifier.predict(X_test_vec)

Evaluasi hasilnya dengan mencetak laporan klasifikasi, matriks kebingungan, dan akurasi:

In [14]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# Cetak laporan klasifikasi
print(classification_report(y_test, y_pred))

# Cetak matriks kebingungan
print(confusion_matrix(y_test, y_pred))

# Cetak akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f'Akurasi: {accuracy}')
print(f'Akurasi (%): {accuracy * 100:.2f}%')

              precision    recall  f1-score   support

         ham       0.99      0.99      0.99       965
        spam       0.96      0.92      0.94       150

    accuracy                           0.98      1115
   macro avg       0.97      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115

[[959   6]
 [ 12 138]]
Akurasi: 0.9838565022421525
Akurasi (%): 98.39%


### 3. Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,
- Menggunakan data spam.csv
- Fitur TF-IDF dengan mengaktifkan stop_words
- Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.
- Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv

In [16]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# Membaca data dari file CSV
dpath = 'dataset/spam.csv'
data = pd.read_csv(dpath, encoding='latin-1') # spesifiksi encoding diperlukan karena data tidak menggunakan UTF-8
data.head()

# Pisahkan atribut (X) dan label (y)
X = data['v2']
y = data['v1']

# Bagi data menjadi data pelatihan dan data pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Buat TfidfVectorizer dengan mengaktifkan stop words:

In [17]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Buat objek TfidfVectorizer dengan pengaktifan stop_words
tfidf_vectorizer = TfidfVectorizer(stop_words='english')

# Transformasikan data teks pelatihan dan pengujian menjadi matriks TF-IDF
X_train_tfidf = tfidf_vectorizer.fit_transform(X_train)
X_test_tfidf = tfidf_vectorizer.transform(X_test)

Buat model Multinomial Naive Bayes dan latih model:

In [18]:
from sklearn.naive_bayes import MultinomialNB

# Buat model Multinomial Naive Bayes
nb_classifier = MultinomialNB()

# Latih model dengan data pelatihan yang telah di-TF-IDF
nb_classifier.fit(X_train_tfidf, y_train)

MultinomialNB()

Lakukan prediksi dengan model yang telah dilatih:

In [19]:
# Lakukan prediksi pada data pengujian yang telah di-TF-IDF
y_pred_tfidf = nb_classifier.predict(X_test_tfidf)


     
Evaluasi hasilnya dengan mencetak laporan klasifikasi, matriks kebingungan, dan akurasi:

In [20]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# Cetak laporan klasifikasi
print("Hasil dengan TF-IDF:")
print(classification_report(y_test, y_pred_tfidf))

# Cetak matriks kebingungan
print("Matriks Kebingungan dengan TF-IDF:")
print(confusion_matrix(y_test, y_pred_tfidf))

# Cetak akurasi
accuracy_tfidf = accuracy_score(y_test, y_pred_tfidf)
print(f'Akurasi dengan TF-IDF: {accuracy_tfidf}')
print(f'Akurasi dengan TF-IDF (%): {accuracy_tfidf * 100:.2f}%')

Hasil dengan TF-IDF:
              precision    recall  f1-score   support

         ham       0.96      1.00      0.98       965
        spam       1.00      0.75      0.86       150

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.92      1115
weighted avg       0.97      0.97      0.96      1115

Matriks Kebingungan dengan TF-IDF:
[[965   0]
 [ 37 113]]
Akurasi dengan TF-IDF: 0.9668161434977578
Akurasi dengan TF-IDF (%): 96.68%


### Kesimpulan<br>
Jadi, dalam kasus dataset 'spam.csv' ini, fitur CountVectorizer dapat dianggap sebagai fitur yang lebih baik digunakan dalam membangun model klasifikasi, karena model klasifikasi yang menggunakan Fitur CountVectorizer dengan pengaktifan stop_words menghasilkan akurasi yang lebih tinggi dengan skor 98.39%, sedangkan model klasifikasi yang menggunakan Fitur TF-IDF dengan pengaktifan stop_words menghasilkan akurasi yang lebih rendah dengan skor 96.68%.